### PR TITLE
Adds the KWN model to NEML2

### PR DESCRIPTION
**Pull request by reverendbedford**
Adds the Kampmann–Wagner Numerical (KWN) model of nucleation, growth, and coarsening to NEML2 as a new physics module.  This represents the nucleation, growth, and coarsening of precipitates by evolving the number density distribution of the precipitate population over time.  The new `kwn` module relies on the `transport` module to solve the resulting advection-reaction equations.

### CODE REVIEW

File: src/neml2/models/finite_volume/DumpInSmallestBin.cxx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+  const auto N = _cell_centers.intmd_size(-1);
+  const auto tail = _cell_centers.intmd_slice(-1, Slice(1, N));
+  const auto mag_raw = _magnitude();
+  const auto mag = (_magnitude.intmd_dim() == 0 ? Scalar(mag_raw.unsqueeze(mag_raw.dynamic_dim()),
+                                                         /*batch_dim=*/mag_raw.dynamic_dim(),
+                                                         /*intmd_dim=*/1)
+                                                : mag_raw.intmd_slice(-1, Slice(0, 1)));
+  const auto zero_tail = Scalar::zeros_like(tail).dynamic_expand_as(mag);
+  const auto source = intmd_cat({mag, zero_tail}, 0);
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply src/neml2/models/finite_volume/DumpInSmallestBin.cxx:74]: What are you talking about?  The left-most bin is always smallest.

File: doc/content/modules/kwn.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+A complete KWN model in NEML2 is composed by combining the precipitation building blocks below with the [finite volume](@ref finite-volume) module, which provides the spatial discretization of the population balance equation for the number density distribution of the precipitates as a function of their radius.
+
+## Theory
+
+The state of the system is the precipitate number density per unit volume, \f$n_i\f$, in each radius bin \f$\left[R_{i-1/2}, R_{i+1/2}\right]\f$ with center \f$R_i\f$. Concentrations are stored as mole fractions \f$x_i\f$. The matrix is treated as a homogeneous reservoir whose composition is determined by mass conservation against the precipitates.
```
Comment (hugary1995) [inline doc/content/modules/kwn.md]: would be nice to add a few words on the indexing, something like $i \in {0, 1, ..., N}$ where $N$ is the number of bins.

Comment (hugary1995) [reply doc/content/modules/kwn.md]: is it useful to also explain the assumption that the precipitate radius is assumed/approximated to be constant in each bin, or is that too much detail?

File: doc/content/modules/kwn.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+### Mass conservation
+
+The volume fraction of a given precipitate population is obtained by integrating the size distribution,
+
+\f[
+  f = \sum_i \frac{4}{3}\pi R_i^3 n_i,
```
Comment (hugary1995) [inline doc/content/modules/kwn.md]: i think this follows the constant-radius-per-bin assumption.

File: doc/content/modules/kwn.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+\f]
+
+and the current matrix concentration of species \f$i\f$ follows from the closure that everything not locked up in precipitates must be in solution,
+
+\f[
+  x^\infty_i = \frac{x_{0,i} - \sum_k f_k x^p_{k,i}}{1 - \sum_k f_k},
```
Comment (hugary1995) [inline doc/content/modules/kwn.md]: we use $i$ for both bin index and species index, which can be very misleading.

File: doc/content/modules/kwn.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+\f]
+
+where \f$x_{0,i}\f$ is the initial concentration, \f$f_k\f$ is the volume fraction of the \f$k\f$-th precipitate, and \f$x^p_{k,i}\f$ is the concentration of species \f$i\f$ in that precipitate.
```
Comment (hugary1995) [inline doc/content/modules/kwn.md]: > where \f$x_{0,i}\f$ is the initial concentration

this would only hold if precipitate density is zero at the beginning. i think a more rigorous statement would be $x_{0,i}$ is the total concentration of species $i$.

Comment (hugary1995) [reply doc/content/modules/kwn.md]: > \f$x^p_{k,i}\f$ is the concentration of species \f$i\f$ in that precipitate

in precipitate $k$

File: doc/content/modules/kwn.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+### Population balance
+
+Each precipitate population evolves according to
+
+\f[
+  \frac{\partial n}{\partial t} + \frac{\partial \left(n \dot{R}\right)}{\partial R} = J_{\text{nuc}},
```
Comment (hugary1995) [inline doc/content/modules/kwn.md]: is $J_{\text{nuc}}$ a function of R?

File: doc/content/modules/kwn.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+\f]
+
+a 1D advection equation in radius space with a nucleation source on the right-hand side. Spatial discretization is provided by the [finite volume](@ref finite-volume) module: cells are defined on \f$R\f$, the growth rate \f$\dot{R}\f$ becomes the advection velocity at cell edges, and the nucleation source is dumped into the smallest bin near the critical radius via `DumpInSmallestBin`.
```
[the remark this answers is not in the document]
Comment (hugary1995) [reply doc/content/modules/kwn.md]: lol i was about to make the same comment as copilot

File: doc/content/modules/kwn.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+### Growth rate
+
+Two growth-rate models are provided.
+
+The simpler **rate-limited** form solves for a single rate-limiting species \f$\eta\f$,
+
+\f[
+  \dot{R} = \frac{1}{R} D_\eta \frac{x^\infty_\eta - x^*_\eta}{x^p_\eta - x^*_\eta},
```
Comment (hugary1995) [inline doc/content/modules/kwn.md]: and here it seems like you switched to using $\eta$ as the species index.

Comment (hugary1995) [reply doc/content/modules/kwn.md]: i think it's also important to note here that this form requires we know which species is the "rate-limiting" species $\eta$.

Comment (hugary1995) [reply doc/content/modules/kwn.md]: $D_\eta$ is the diffusion coefficient of species $\eta$ in the host precipitate.

Comment (hugary1995) [reply doc/content/modules/kwn.md]: At this point, I think I prefer carrying the superscript $p$ everywhere to indicate the precipitate indexing, even in the advection-reaction PDE.

File: doc/content/modules/kwn.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+with \f$\Delta x_i = x^p_i - x^*_i\f$. The driving force is split into chemical, surface, and elastic contributions:
+
+\f[
+  \Delta G_{\text{chem}} = \sum_i \Delta x_i \left[\mu^{\text{matrix}}_i(x^\infty) - \mu^{\text{equil}}_i\right], \quad
+  \Delta G_{\text{surf}} = -\frac{2 \gamma V_m}{R}, \quad
+  \Delta G_{\text{el}} = -3 V_m \varepsilon_0 \sigma_h + \frac{6 V_m \mu K}{3K + 4\mu}\varepsilon_0^2.
+\f]
+
+The \f$\Delta G_{\text{el}}\f$ term is the standard expression for a coherent, isotropic spherical inclusion in an isotropic matrix with misfit strain \f$\varepsilon_0\f$.
```
Comment (hugary1995) [inline doc/content/modules/kwn.md]: there are too many symbols left unexplained here

File: doc/content/modules/kwn.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+### Nucleation flux
+
+All nucleation occurs at the critical radius,
```
Comment (hugary1995) [inline doc/content/modules/kwn.md]: is this an implementation choice, model choice, or truth that people should believe?

Comment (hugary1995) [reply doc/content/modules/kwn.md]: would be nice to provide a reference, i.e., is this part of SFFK?

Comment (hugary1995) [reply doc/content/modules/kwn.md]: might also help to explaining what "critical radius" means.

**Review by hugary1995 — CHANGES_REQUESTED** [review]
went through the doco

File: include/neml2/models/kwn/NucleationBarrierandCriticalRadius.h [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+{
+class Scalar;
+
+/// Critical radius and nucleation barrier for nucleation.
+class NucleationBarrierandCriticalRadius : public Model
+{
+public:
+  static OptionSet expected_options();
+
+  NucleationBarrierandCriticalRadius(const OptionSet & options);
```
[the remark this answers is not in the document]
Comment (hugary1995) [reply include/neml2/models/kwn/NucleationBarrierandCriticalRadius.h]: haha I was about to make the same comment

File: include/neml2/models/kwn/SFFKGPrecipitationGrowthRate.h [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+#pragma once
+
+#include "neml2/models/Model.h"
+
+namespace neml2
+{
+class Scalar;
+
+/// SFFK precipitate growth rate for a single precipitate population.
+class SFFKGPrecipitationGrowthRate : public Model
```
Comment (hugary1995) [inline include/neml2/models/kwn/SFFKGPrecipitationGrowthRate.h]: What does G stand for in SFFKG?

File: include/neml2/user_tensors/ProductUserTensor.h [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace neml2
+{
+/**
+ * @brief Compute the product $a b$.
+ *
+ * @tparam T The concrete tensor derived from TensorBase
+ */
+template <typename T>
+class ProductUserTensorTmpl : public UserTensorBase<T>
```
Comment (hugary1995) [inline include/neml2/user_tensors/ProductUserTensor.h:37]: sigh, i should really start working on parsed function

(note for myself)

File: include/neml2/user_tensors/SemiInfiniteScaling.h [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace neml2
+{
+/**
+ * @brief Compute the semi-infinite scaling $s x / (1 - x)$.
+ *
+ * @tparam T The concrete tensor derived from TensorBase
+ */
+template <typename T>
+class SemiInfiniteScalingTmpl : public UserTensorBase<T>
```
Comment (hugary1995) [inline include/neml2/user_tensors/SemiInfiniteScaling.h:37]: I looked at the corresponding implementations of `SemiInfiniteScaling` and `Inver...`, I think the amount of code duplication warrants a merge and the use of a "inverse = true/false" option to branch the behavior.

File: python/examples/kwn/AlCu-Liang-Schmid-Fetzer.tdb
Comment (hugary1995) [inline python/examples/kwn/AlCu-Liang-Schmid-Fetzer.tdb:1]: did the authors provide consent to distribute the database freely?

**Review by hugary1995 — CHANGES_REQUESTED** [review]

File: python/examples/kwn/al-cu-ttp.md [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+### Plot some of the number density distributions
+
+This plot shows the number density distribution from the KWN simulations at different temperatures, for the final time step.  It's hard to follow the colors, but you can see the trend indicated by the TTP diagram: you get the highest volume fraction around 360 K, diminishing at temperatures above and below this critical point.
+
+```{code-cell} ipython3
+edges = torch.logspace(-2,1.25,nbins+1)
+radii = 0.5 * (edges[:-1] + edges[1:])
+plt.semilogx(radii, N_full[::5,-1].T, label = [f"T={T:.0f}K" for T in temperatures[::5]])
+plt.xlabel("Particle radius (mm)")
+plt.ylabel("Number density (#/mm^3)")
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply python/examples/kwn/al-cu-ttp.md:279]: I don't think this is true, it works just fine.

File: doc/content/modules/kwn.md (hunk 6 above)
Comment (reverendbedford) [reply doc/content/modules/kwn.md]: It is kinda, actually only a function of the critical radius.  I now defer this discussion to the nucleation section.

File: doc/content/modules/kwn.md (hunk 2 above)
Comment (reverendbedford) [reply doc/content/modules/kwn.md]: I mean it's the standard finite volume assumption, but I now mention it.

File: doc/content/modules/kwn.md (hunk 5 above)
Comment (reverendbedford) [reply doc/content/modules/kwn.md]: True, added the assumption.

File: doc/content/modules/kwn.md (hunk 9 above)
Comment (reverendbedford) [reply doc/content/modules/kwn.md]: Fixed.

File: doc/content/modules/kwn.md (hunk 10 above)
Comment (reverendbedford) [reply doc/content/modules/kwn.md]: Fixed this.  Nucleation should occur at the critical radius per the classical theory (chemical driving force overcomes surface energy of formation).  However we have to make compromises.  Hopefully the new discussion is more clear.

**Review by hugary1995 — APPROVED** [review]
